Upgrade SB3 and migrate to local logger

AdamGleave opened the pull request:
Move from SB3 0.10.0 to latest version, currently 1.1.0.

SB3 moved from a global logger instance to a local logger for each algorithm. This PR switches to the same style.

I've also tidied up some hacks that we had to make things work with the global logger previously. I'm not sure I've caught all of them though. In particular there were some cases where `verbose=1` was flagged with a FIXME as being a workaround, but `verbose=1` also appears elsewhere in the codebase. I'm not sure whether these were introduced as a workaround or because we actually want that verbosity.

Overall I think this PR should be LOC negative :)

@qxcv: one tricky thing that came up here was that DAgger is pickled to save it (as is, indirectly, BC), but the Logger is not pickleable. I've just removed the Logger from pickle, and setup a default logger on restore. This feels a bit clunky though. I think you've got a better sense than me when DAgger needs to be saved/restored, so if you have a better idea for this let me know.

Fixes #325 and #327.

qxcv commented:
> @qxcv: one tricky thing that came up here was that DAgger is pickled to save it (as is, indirectly, BC), but the Logger is not pickleable. I've just removed the Logger from pickle, and setup a default logger on restore. This feels a bit clunky though. I think you've got a better sense than me when DAgger needs to be saved/restored, so if you have a better idea for this let me know.

That seems fine to me! I guess the only downside is that logs will be stored separately for each restore/save cycle?

AdamGleave commented:
> That seems fine to me! I guess the only downside is that logs will be stored separately for each restore/save cycle?

Yeah, unless the user re-configures the logger to have the same directory at least. (And even then I'm not sure things are safe to re-open?)

We could mess around to make `Logger` pickleable, in terms of e.g. saving the folder and any other state needed to restore (tensorboard step?), but restoration would then break across hosts/filesystems which seems bad in its own way.

ejnnr reviewed (approved):
LGTM except for the small things I commented on (which I don't think require re-review after addressing them).

Two things that probably don't need to be addressed in this PR:
- I agree refactoring to get rid of `utils.init_rl` sounds good (though I haven't checked in detail whether there would be any significant hurdles). Apart from making the code simpler to read, having a CLI argument like `rl_kwargs` or `agent_kwargs` is probably more intuitive than the current `init_rl_kwargs`
- My type checker (Pylance) complains about a couple of type issues and I think it's right about some of them. For example, we should probably use `output_formats=[]` here? https://github.com/HumanCompatibleAI/imitation/blob/dd1b747714fc42d44261931fc88b1022ce23d296/src/imitation/util/logger.py#L51
  And there are a couple of places where we should technically convert an `AnyPath` to `str` before passing it on. But I think all of them already existed before the PR anyway.

AdamGleave commented on src/imitation/util/logger.py:
```diff
@@ -160,55 +144,15 @@ def configure(
-    sb_logger.configure = _sb_logger_configure_replacement
-    sb_logger.reset = _sb_logger_reset_replacement
-
+    if folder is None:
+        if folder is None:
+            now = datetime.datetime.now()
+            timestamp = now.strftime("imitation-%Y-%m-%d-%H-%M-%S-%f")
+            folder = os.path.join(tempfile.gettempdir(), timestamp)
```
Doh, fixed.

AdamGleave commented on src/imitation/algorithms/bc.py:
```diff
@@ -414,3 +418,14 @@ def save_policy(self, policy_path: types.AnyPath) -> None:
+    def __setstate__(self, state):
+        self.__dict__.update(state)
+        # callee should call set_logger if they want to override this
+        self._logger = state.get("logger") or logger.configure()
```
oops, yes

AdamGleave commented on setup.py:
```diff
@@ -56,7 +56,7 @@ def get_readme() -> str:
         "tqdm",
         "scikit-learn>=0.21.2",
-        "stable-baselines3~=0.10.0",
+        "stable-baselines3",
```
Good point, changed.

AdamGleave commented on src/imitation/algorithms/dagger.py:
```diff
@@ -314,12 +315,29 @@ def __init__(
         self._last_loaded_round = -1
         self._all_demos = []
 
+        self._logger = custom_logger or logger.configure()
```
It is deliberate. We need to set the BC logger from DAgger so I actually want `logger` exposed for BC. Granted the inconsistency is grating, so maybe we should go with `_logger` and `set_logger` for both, but on balance I think additional complexity isn't worth it.

AdamGleave commented:
Thanks for the review!

I've opened https://github.com/HumanCompatibleAI/imitation/issues/327 to track `init_rl` removal.

Agree there are some type errors. I've introduced a new function to convert `AnyPath` to `str`.

Also added some missing functions in `Logger` and expanded test coverage.
